subworkflows/local/prepare_genome/main.nf

```diff
@@ -0,0 +1,149 @@
+    ch_trna                   = val_trna                      ? Channel.fromPath(val_trna, checkIfExists: true).map{ it -> [ [id:'tRNA'], it ] }.collect()                 : Channel.empty()
+    ch_cdna                   = val_cdna                      ? Channel.fromPath(val_cdna, checkIfExists: true).map{ it -> [ [id:'cDNA'], it ] }.collect()                 : Channel.empty()
+    ch_ncrna                  = val_ncrna                     ? Channel.fromPath(val_ncrna, checkIfExists: true).map{ it -> [ [id:'ncRNA'], it ] }.collect()               : Channel.empty()
+    ch_pirna                  = val_pirna                     ? Channel.fromPath(val_pirna, checkIfExists: true).map{ it -> [ [id:'piRNA'], it ] }.collect()               : Channel.empty()
+    ch_other_contamination    = val_other_contamination       ? Channel.fromPath(val_other_contamination, checkIfExists: true).map{ it -> [ [id:'other'], it ] }.collect() : Channel.empty()
+
+    // even if bowtie index is specified, there still needs to be a fasta.
+    // without fasta, no genome analysis.
+    if(val_fasta) {
+        // Clean fasta (replace non-ATCGs with Ns, remove whitespaces from headers)
+        // Note: CLEAN_FASTA runs even when a bowtie_index is provided, as cleaning doesn't affect it, making regeneration unnecessary.
+        CLEAN_FASTA ( ch_fasta )
+        ch_versions      = ch_versions.mix(CLEAN_FASTA.out.versions)
+        ch_fasta         = CLEAN_FASTA.out.output
+
+        //Prepare bowtie index, unless specified
+        //This needs to be done here as the index is used by GENOME_QUANT
+        if(val_bowtie_index) {
+            if (val_bowtie_index.endsWith(".tar.gz")) {
+                UNTAR_BOWTIE_INDEX ( ch_bowtie_index )
+                ch_bowtie_index = UNTAR_BOWTIE_INDEX.out.untar
+                    .map{ meta, index_dir ->
+                        def index_prefix = extractFirstIndexPrefix(index_dir)
+                        [[id:index_prefix], index_dir]
+                    }
+                ch_versions  = ch_versions.mix(UNTAR_BOWTIE_INDEX.out.versions)
+            } else {
+                ch_bowtie_index = Channel.fromPath(val_bowtie_index, checkIfExists: true)
+                    .map{it ->
+                        def index_prefix = extractFirstIndexPrefix(it)
+                        [[id:index_prefix], it]
+                    }
+            }
+
+        } else {
+
+            // Index FASTA with nf-core Bowtie1
+            INDEX_GENOME ( CLEAN_FASTA.out.output )
+            ch_versions      = ch_versions.mix(INDEX_GENOME.out.versions)
+
+            // Set channels: clean fasta and its index
+            ch_bowtie_index  = INDEX_GENOME.out.index.collect()
+        }
+    }
+
+    //Config checks
+    // Check optional parameters
+    if (!params.mirgenedb && !val_mirtrace_species) {
```

Comment:
I would centralise all input validation to `utils_nfcore_smrnaseq_pipeline` in the `validateInputParameters` function, if you have incompatible parameters you wqant to fail ASAP (assuming the invalid parameters do not require some for of preprocessing)

Reply:
Made an issue out of this for the next iteration.

Reply:
Hi @nschcolnicov, I was wondering if there was any specific reason you decided to place the input validation in `prepare_genome` rather than in `utils_nfcore_*`?